cloudferrylib/utils/utils.py

```diff
@@ -296,27 +296,45 @@ def inner(*args, **kwargs):
     return decorator
 
 
-class forward_agent:
-
+class forward_agent(object):
     """
         Forwarding ssh-key for access on to source and destination clouds via ssh
     """
 
     def __init__(self, key_file):
         self.key_file = key_file
+        self.agent_is_already_running = False
+
+    def _agent_already_running(self):
+        with settings(hide('warnings', 'running', 'stdout', 'stderr'),
+                      warn_only=True):
+            res = local("ssh-add -l", capture=True)
+
+            if res.succeeded:
+                present_keys = res.split(os.linesep)
+                for key in present_keys:
+                    # TODO: this will break for path with whitespaces
+                    key_path = key.split(' ')[2]
+                    if key_path == os.path.expanduser(self.key_file):
+                        self.agent_is_already_running = True
+
+        return self.agent_is_already_running
 
     def __enter__(self):
-        info_agent = local("eval `ssh-agent` && echo $SSH_AUTH_SOCK && ssh-add %s" %
-                           (self.key_file), capture=True).split("\n")
+        if self._agent_already_running():
+            return
+        start_ssh_agent = ("eval `ssh-agent` && echo $SSH_AUTH_SOCK && "
+                           "ssh-add %s") % self.key_file
+        info_agent = local(start_ssh_agent, capture=True).split("\n")
         self.pid = info_agent[0].split(" ")[-1]
         self.ssh_auth_sock = info_agent[1]
         os.environ["SSH_AGENT_PID"] = self.pid
         os.environ["SSH_AUTH_SOCK"] = self.ssh_auth_sock
 
     def __exit__(self, type, value, traceback):
-        local("kill -9 %s"%(self.pid))
-        del os.environ["SSH_AGENT_PID"]
-        del os.environ["SSH_AUTH_SOCK"]
+        # never kill previously started ssh-agent, so that user only has to
+        # enter private key password once
```

Comment:
Yes, exactly that. And I don't see any problem with that. Actually I think ssh-agent should be started prior to running cloudferry.